crates/matrix-sdk/src/widget_api/driver/widget.rs

```diff
@@ -0,0 +1,78 @@
+use async_trait::async_trait;
+
+use super::handler::Result;
+use crate::widget_api::messages::capabilities;
+
+/// The Widget trait that needs to be implemented in the native client.
+/// An instance implementing this trait will be required to initialize the widget driver.
+#[async_trait]
+pub trait Widget {
+    /// This function will be called whenever there is a new message available
+    /// that has to be sent to the widget.
+    ///
+    /// # Examples
+    /// ```
+    /// implement Widget for MyWidget {
+    ///     fn send_widget_message(&self, message: &str) -> Result<()> {
+    ///         myWidgetIFrame.postmessage(message);
+    ///         Ok()
+    ///     }
+    /// }
+    ///```
+    fn send_widget_message(&self, message: &str) -> Result<()>;
+
+    /// The client should show a dialog to give the user the option approve some/all of the capabilites that the widget requests.
+    /// The returned value contains the capability options the user has approved.
+    /// The client should also provide good phrasing for the different permissions/filters:
+    /// cap.send_room_event = [{event_type: "m.room.message", msgtype: "m.image" }] should be come sth like:
+    /// Allow the widget to, send images in this room. (checkbox)
+    async fn capability_permissions(
+        &self,
+        cap: capabilities::Options,
+    ) -> Result<capabilities::Options>;
+
+    /// The client should show a dialog to approve if the widget is allowed to get an OpenId token.
+    /// A `OpenIdDialogResponse` is returned containing a flag if the user alloed the token request and
+    /// if the permission can be cached so the widget driver will not call `openid_permissions` on the widget again.
+    async fn openid_permissions(&self) -> Result<OpenIdDialogResponse>;
```

Comment:
this goes with the same topic of capabilites beeing seperate from the open id logic. (historical resons we talked about before)
So capabiliteis will be asked for in the beginning but there is not way for the widget to communicate that it also wants to have capabiliteis for the open id things. Therefore they will never be called at the same time. That is why i think two functions make sense?

Reply:
But isn't it part of the widget's/client's state? I believe it will have a state saved anyway, so it would be a single function at the end. Otherwise this state of "already requested vs not yet requested" will have to live somewhere else.